Replace attack() with use_before()

emmanuelbassil opened the pull request:
🆑 emmanuelbassil
tweak: Clicking on yourself with a pill bottle no matter the targeted area consumes a pill. Previous behavior required targeting mouth.
/🆑 

Depends on #34157.
Replaced attack() with use_before() and use_on() with use_after(). Removed the attack override flag and incorporated both of these in the resolve_attackby() chain.

While testing, noticed pill_bottle had two separate afterattacks() defined, so the pill popping behavior was not working. Also noticed they're a good candidate to be switched to use_after, and did so. The only afterattacks() I will have trouble switching to use_after are those few that act in a RANGED way.

emmanuelbassil commented:
Of note:
The issues I had with welding/ATKs/nanopaste in surgery that were fixed in another open PR can also be properly mitigated by moving alot of what’s under attack() (now use_before) to use_after.

The early override should be limited to behavior one actually wants to override/come before use_weapon and use_tool. Will look into this in future PRs

emmanuelbassil committed:
Fixes popping pills

Fixes

Auto stash before rebase of "refs/heads/paddingfix"

emmanuelbassil committed:
Replaces attack() with use_before()

Fixes

emmanuelbassil commented:
Fixed merge conflicts.

SierraKomodo reviewed (commented):
Nothing immediately stands out to me, should be fine to do a test run at some point.

We have a lot of test merge flagged PRs, good lord.

emmanuelbassil commented:
This is ready, but I still need to locally test the changes of the last commit.

Changed some use_befores to use_afters. Did not do all the switches I could theoretically do, if it ain't broke don't fix it. The bulk of use_afters will come from afterattack in separate PRs. Given I want to avoid more test merges, those will be piecemeal. 
Chose not to change the surgical use_befores to use_afters, meaning I'm keeping the check_possible_surgeries proc for now.

SierraKomodo commented on code/game/objects/items/weapons/tools/weldingtool.dm:
```diff
@@ -280,7 +279,8 @@
 
-/obj/item/weldingtool/attack(mob/living/M, mob/living/user)
+/obj/item/weldingtool/use_before(mob/living/M, mob/living/user)
+	. = FALSE
```
Unnecessary since you're returning values anyway.

```suggestion
```

SierraKomodo commented on code/modules/clothing/shoes/miscellaneous.dm:
```diff
@@ -230,16 +230,21 @@
-		playsound(src.loc, 'sound/items/Wirecutter.ogg', 100, 1)
-		user.visible_message(SPAN_WARNING("\The [user] modifies \the [src] with \the [W]."),SPAN_WARNING("You modify \the [src] with \the [W]."))
-		cut_clogs()
+	if (!isitem(W))
+		return FALSE
+	if (!is_sharp(W))
```
`isitem()` checks are redundant here.

```suggestion
	if (!is_sharp(W))
```

SierraKomodo commented on code/modules/clothing/_clothing.dm:
```diff
@@ -207,6 +207,16 @@
 
+/obj/item/clothing/use_tool(obj/item/tool, mob/living/user, list/click_params)
+	SHOULD_CALL_PARENT(TRUE)
+	if (!isitem(tool))
+		return FALSE
+	if (attempt_attach_accessory(tool, user))
```
```suggestion
	if (attempt_attach_accessory(tool, user))
```

emmanuelbassil commented:
This is ready again; depends on the giftwrapper PR.

Made the  surgical overrides use_after. As a result, I could safely deleted the check_possible_surgeries proc. Only downside is that using a welding tool on a robotic hand while the person is on an OR table will return the failed surgical step. I could add an intent dependent behavior (like for advanced trauma kits) but will leave this for another PR as maybe robotics players don't want such a change.

Also made gift_wrap a use_after; meaning no need for checks to allow base stack functionality to work.
Made a few other use_after's, but didn't as much as I could've. If it ain't broke, don't fix it.

One thing: For the use_calls that are used by post_use_item; I elected to have use_before and use_after both have the "use" label. I felt having a separate one for before and after is unnecessary and would add more convolution.

SierraKomodo reviewed (dismissed):
Merge conflicts

emmanuelbassil committed:
Switches some befores to afters

Fixes

Auto stash before rebase of "refs/heads/use_before/after"

Fixes

Auto stash before rebase of "refs/heads/QIAssort"
